api/file.go

```diff
@@ -103,82 +105,92 @@ func uploadFile(c *Context, w http.ResponseWriter, r *http.Request) {
 
-		uid := model.NewId()
+		filename := filepath.Base(fileHeader.Filename)
+		pathPrefix := "teams/" + c.TeamId + "/channels/" + channelId + "/users/" + c.Session.UserId + "/" + info.Id + "/"
 
-		if model.IsFileExtImage(filepath.Ext(files[i].Filename)) {
-			imageNameList = append(imageNameList, uid+"/"+filename)
-			imageDataList = append(imageDataList, buf.Bytes())
+		info.Path = pathPrefix + filename
 
-			// Decode image config first to check dimensions before loading the whole thing into memory later on
-			config, _, err := image.DecodeConfig(bytes.NewReader(buf.Bytes()))
-			if err != nil {
-				c.Err = model.NewLocAppError("uploadFile", "api.file.upload_file.image.app_error", nil, err.Error())
-				return
-			} else if config.Width*config.Height > MaxImageSize {
+		if info.IsImage() {
+			// Check dimensions before loading the whole thing into memory later on
+			if info.Width*info.Height > MaxImageSize {
 				c.Err = model.NewLocAppError("uploadFile", "api.file.upload_file.large_image.app_error", nil, c.T("api.file.file_upload.exceeds"))
+				c.Err.StatusCode = http.StatusBadRequest
 				return
 			}
-		}
 
-		path := "teams/" + c.TeamId + "/channels/" + channelId + "/users/" + c.Session.UserId + "/" + uid + "/" + filename
+			nameWithoutExtension := filename[:strings.LastIndex(filename, ".")]
+			info.PreviewPath = pathPrefix + nameWithoutExtension + "_preview.jpg"
+			info.ThumbnailPath = pathPrefix + nameWithoutExtension + "_thumb.jpg"
+			previewPathList = append(previewPathList, info.PreviewPath)
+			thumbnailPathList = append(thumbnailPathList, info.ThumbnailPath)
+			imageDataList = append(imageDataList, buf.Bytes())
+		}
 
-		if err := WriteFile(buf.Bytes(), path); err != nil {
+		if err := WriteFile(buf.Bytes(), info.Path); err != nil {
 			c.Err = err
 			return
 		}
 
-		encName := utils.UrlEncode(filename)
+		if result := <-Srv.Store.FileInfo().Save(info); result.Err != nil {
+			c.Err = result.Err
+			return
+		}
 
-		fileUrl := "/" + channelId + "/" + c.Session.UserId + "/" + uid + "/" + encName
-		resStruct.Filenames = append(resStruct.Filenames, fileUrl)
-	}
+		resStruct.FileInfos = append(resStruct.FileInfos, info)
 
-	for _, clientId := range props["client_ids"] {
-		resStruct.ClientIds = append(resStruct.ClientIds, clientId)
+		if len(m.Value["client_ids"]) > 0 {
+			resStruct.ClientIds = append(resStruct.ClientIds, m.Value["client_ids"][i])
+		}
 	}
 
-	go handleImages(imageNameList, imageDataList, c.TeamId, channelId, c.Session.UserId)
+	go handleImages(previewPathList, thumbnailPathList, imageDataList)
 
 	w.Write([]byte(resStruct.ToJson()))
 }
 
-func handleImages(filenames []string, fileData [][]byte, teamId, channelId, userId string) {
-	dest := "teams/" + teamId + "/channels/" + channelId + "/users/" + userId + "/"
-
-	for i, filename := range filenames {
-		name := filename[:strings.LastIndex(filename, ".")]
+func handleImages(previewPathList []string, thumbnailPathList []string, fileData [][]byte) {
```

Comment:
Yeah, I'll take a look at it. It's already kind of weird since we call handleImages in a goroutine and then it spins up goroutines that spin up goroutines.